mule-shaders-module/src/main/java/org/mule/module/shaders/api/ReflectiveAccessShader.java

```diff
@@ -0,0 +1,641 @@
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.shade.ShadeRequest;
+import org.apache.maven.plugins.shade.Shader;
+import org.apache.maven.plugins.shade.filter.Filter;
+import org.apache.maven.plugins.shade.relocation.Relocator;
+import org.apache.maven.plugins.shade.resource.ManifestResourceTransformer;
+import org.apache.maven.plugins.shade.resource.ReproducibleResourceTransformer;
+import org.apache.maven.plugins.shade.resource.ResourceTransformer;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.logging.AbstractLogEnabled;
+import org.codehaus.plexus.util.IOUtil;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.commons.Remapper;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.PushbackInputStream;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+
+/**
+ *  A shader capable of changing the specified classes for reflective access before relocating packages.
+ *  Custom version of {@link org.apache.maven.plugins.shade.DefaultShader}.
+ */
+@Component(role = Shader.class, hint = "ReflectiveAccessShader")
+public class ReflectiveAccessShader
+    extends AbstractLogEnabled
+    implements Shader {
+
+  private static final int BUFFER_SIZE = 32 * 1024;
+
+  public void shade(ShadeRequest shadeRequest)
+      throws IOException, MojoExecutionException {
+    Set<String> resources = new HashSet<>();
+
+    ManifestResourceTransformer manifestTransformer = null;
+    List<ResourceTransformer> transformers =
+        new ArrayList<>(shadeRequest.getResourceTransformers());
+    for (Iterator<ResourceTransformer> it = transformers.iterator(); it.hasNext();) {
+      ResourceTransformer transformer = it.next();
+      if (transformer instanceof ManifestResourceTransformer) {
+        manifestTransformer = (ManifestResourceTransformer) transformer;
+        it.remove();
+      }
+    }
+
+    RelocatorRemapper remapper = new RelocatorRemapper(shadeRequest.getRelocators());
+
+    // noinspection ResultOfMethodCallIgnored
+    shadeRequest.getUberJar().getParentFile().mkdirs();
+
+    JarOutputStream out = null;
+    try {
```

Comment:
As the class javadoc points out, `ReflectiveAccessShader` is basically a custom version of `org.apache.maven.plugins.shade.DefaultShader`, and it was copied over from that code because we needed to modify one line of code in order to inject our own class remapper, see discussions on  https://github.com/mulesoft/mule-json-module/pull/196.

While we might change to code to make it cleaner, there is no functional reason to change it because this is how the default shade plugin behaves anyway, and the next time we have to make a version upgrade for that plugin, any changes we do would just make that process more complicated.